Move redis adapter to separate npm library

mannyv123 opened the pull request:
### Description
This change removes the redis 4 adapter from `@epic-web/cachified` and links to a separate npm library I set up:
[cachified-redis-adapter](https://github.com/mannyv123/cachified-redis-adapter)

### Why
This helps avoid issues when updating the adapter and causing breaking change releases for the entire `@epic-web/cachified` package.

### Details
- `cachified-redis-adapter` basically takes the code that was in `@epic-web/cachified`  for the redis adapter and moves it into a separate library
- this PR then updates `@epic-web/cachified` to remove the redis adapter code and update the `readme.md` file to link to `cachified-redis-adapter`


**Note**: I wasn't too sure how to update the `extractExamples.js` file. I can see that since the `readme.md` file is update it has removed the redis adapter example from generating. Any suggestions here would be appreciated!

@tearingItUp786

Xiphe reviewed (commented):
Code-wise: LGTM

But we need to decide if we want to release this as a breaking change or not.

- A) When we want to actually remove the adapter:   
  Add `BREAKING CHANGE` to one of the commit messages (or to the squash merge message) in order to issue a new major version. See [semantic-release#commit-message-format](https://github.com/semantic-release/semantic-release?tab=readme-ov-file#commit-message-format) for details.
- B) When we don't want to create a new major version just yet:  
  Keep the adapter in and `@deprecate`.  The docs can be updated to point towards the new package.

Leaving this decision to you @kentcdodds

mannyv123 commented:
Thanks everyone for the feedback! I've made the following changes:
- removed redis3 adapter code and related tests
- removed `redis-mock`, `@types/redis-mock`, `redis4` packages

I also added `BREAKING CHANGE` to the commit message, hopefully formatted it correctly🤞

Thanks again!

kentcdodds reviewed (changes requested):
Actually, the more I think about it, the more I think I would prefer to delete all adapters. And I think the README could be cleaned up as well. Do you have time and interest in doing that please @mannyv123?

tearingItUp786 commented:
Hey hey! I'll probably be helping @mannyv123 with this 😄 

So the plan for the work going forward: 
- remove `adapters.ts` 
- Update the `README`
- Update the `extractExamples.js` or do we want to remove this?

kentcdodds commented:
I think we can remove that. Thanks @tearingItUp786!

Xiphe commented:
Voting to keep `extractExamples.js` but as the creator of it I'm certainly biased :)

I've created that one because I tend to forget keeping (readme) examples up to date and wanted to have a safeguard. 
Let me know when you want my help with it.

kentcdodds commented:
I think it's a cool idea, but with removing all the adapters and moving the docs for them into their respective packages combined with the level of complexity there, I don't think it makes sense to keep it around.

Xiphe commented:
The adapters were actually the examples that have not been tested :) 
But I'm completely fine with removing it 👍

mannyv123 commented:
Hi all!

I've made the changes as discussed:
- removed adapters and tests
- removed `extractExamples.js` and related tests
- updated the `README`

@kentcdodds you mentioned during the last office hours that it would be good to update the usage example in the `README` with the lru adapter code and that we might be able to reduce the amount of code as well. I updated the example with the lru adapter but wasn't too sure how to make it simpler. Would you be able to provide some feedback/guidance on how to update this? Thank you!

kentcdodds commented on README.md:
```diff
@@ -52,10 +52,40 @@ npm install @epic-web/cachified
-const lru = new LRUCache<string, CacheEntry>({ max: 1000 });
+const lruInstance = new LRUCache<string, CacheEntry>({ max: 1000 });
+
+/* defines options for the set method for lru cache */
+interface LRUishCache extends Omit<Cache, 'set'> {
+  set(
+    key: string,
+    value: CacheEntry<unknown>,
+    options?: { ttl?: number; start?: number },
+  ): void;
+}
+
+/* creates a wrapper for the lru cache so that it can easily work with cachified and ensures the lru cache cleans up outdated values itself*/
+function lruCacheAdapter(lruCache: LRUishCache): Cache {
+  return {
+    set(key, value) {
+      const ttl = totalTtl(value?.metadata);
+      return lruCache.set(key, value, {
+        ttl: ttl === Infinity ? undefined : ttl,
+        start: value?.metadata?.createdTime,
+      });
+    },
+    get(key) {
+      return lruCache.get(key);
+    },
+    delete(key) {
+      return lruCache.delete(key);
+    },
+  };
+}
+
+const lru = lruCacheAdapter(lruInstance);
```
Remove this stuff and just set `lru` to the cache instance itself. So something like:

```
const lru: Cache = {
  set() {},
  get() {},
  delete() {},
}
```

Of course you'll want to fill that in with working code and make sure the types are happy too. I think you'll want to use the Cache generic maybe.

kentcdodds reviewed (commented):
This is looking good to me. Definitely simplifies things for us development-wise and reduces the chance that we'll need breaking changes in the future.

Just one thing on the example in the README and then I'll be ready to merge this.

Xiphe reviewed (approved):
Agree with Kents feedback. Rest: LGTM!

mannyv123 commented:
Thank you for the feedback!

I've updated the usage example now. Please let me know if this is in line with what you were thinking or if any changes are needed. 

Thanks again!

Xiphe commented on README.md:
```diff
@@ -52,40 +52,23 @@ npm install @epic-web/cachified
+  delete(key) {
+    return lruInstance.delete(key);
+  },
+};
```
This implementation has no benefit over

```ts
const cache = new LRUCache<string, CacheEntry>({ max: 1000 })

cachified({ cache, /* ... */ });
```

One of the main reasons for an adapter is to set the TTL on the cached item in a way that the cache will clean it up itself without involving `cachified`.

Proposing to either not create an on the fly adapter (example above) or do the following:

```ts
/* lru cache is not part of this package but a simple non-persistent cache */
const lruInstance = new LRUCache<string, CacheEntry>({ max: 1000 });

const cache: Cache = {
  set(key, value) {
    const ttl = totalTtl(value?.metadata);
    return lruCache.set(key, value, {
      ttl: ttl === Infinity ? undefined : ttl,
      start: value?.metadata?.createdTime,
    });
  },
  get(key) {
    return lruInstance.get(key);
  },
  delete(key) {
    return lruInstance.delete(key);
  },
};
```

Xiphe replied:
Would vote for the latter option. In real apps this is quite critical in order to not bloat your cache db.

mannyv123 replied:
Ahh that makes sense! Thank you for the help!

I've updated it now to the above example.

mannyv123 commented:
Hi all,

I resolved the merge conflicts that were in the package.json file. 

Also, I made an update to the `tsconfig.json` file as when `tsc` was run, it was no longer creating the `dist/src` folder structure. It seems this was caused when I removed the `readme.test.ts` file. Please let me know if my change here was incorrect and I can update.

Thanks!

Xiphe reviewed (approved):
LGTM :whale2:

Good catch with the `src` folder. The other option would have been to remove the `src/` in the `"type"` paths in package.json but I'm fine with either 👍

caiocmpaes commented:
Hi, everyone! 👋 

I just updated the package to `5.1.1` and then seems like we need a bump on the peer dependency of [cachified-redis-adapter](https://www.npmjs.com/package/cachified-redis-adapter).

<img width="791" alt="Captura de Tela 2024-02-09 às 21 48 04" src="https://github.com/epicweb-dev/cachified/assets/14334253/d0d19097-4b7a-49c8-8886-0f54419170af">

tearingItUp786 commented:
> Hi, everyone! 👋
> 
> I just updated the package to `5.1.1` and then seems like we need a bump on the peer dependency of [cachified-redis-adapter](https://www.npmjs.com/package/cachified-redis-adapter).
> 
> <img alt="Captura de Tela 2024-02-09 às 21 48 04" width="791" src="https://private-user-images.githubusercontent.com/14334253/303804299-d0d19097-4b7a-49c8-8886-0f54419170af.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc1MzkyNjAsIm5iZiI6MTcwNzUzODk2MCwicGF0aCI6Ii8xNDMzNDI1My8zMDM4MDQyOTktZDBkMTkwOTctNGI3YS00OWM4LTg4ODYtMGY1NDQxOTE3MGFmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjEwVDA0MjI0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFhOGYyYWI4ZmEwMDFmZGVkYzA5NTI2ZTVhNTNlMzJlZDI3NDU5YjBhZDA0ZDU2MjA5NGNjOGRhYmYzNDM3OTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Fv0rIAA4q41nGkPDRS8xUT-FTnIQyk6--lhoiXJgc6s">

https://github.com/mannyv123/cachified-redis-adapter/pull/7

mannyv123 commented:
@caiocmpaes @tearingItUp786 PR is now merged! :)
Thank you!
